src/gt4py/next/otf/workflow.py

```diff
@@ -128,3 +128,41 @@ def __call__(self, inp: StartT) -> EndT:
 
     def chain(self, step: Workflow[EndT, NewEndT]) -> CombinedStep[StartT, EndT, NewEndT]:
         return CombinedStep(first=self, second=step)
+
+
+@dataclasses.dataclass(frozen=True)
+class CachedWorkflow(Generic[StartT, EndT]):
+    """
+    Cached workflow of single input callables.
+
+    Examples:
+    ---------
+    >>> def heavy_computation(x: int) -> int:
+    ...    print("This might take a while...")
+    ...    return x
+
+    >>> cached_workflow = CachedWorkflow(heavy_computation)
+
+    >>> cached_workflow(42)
+    This might take a while...
+    42
+
+    The next invocation for the same argument will be cached:
+    >>> cached_workflow(42)
+    42
+
+    >>> cached_workflow(1)
+    This might take a while...
+    1
+    """
+
+    workflow: Workflow[StartT, EndT]
+    hash_function: Callable[[StartT], Any] = dataclasses.field(default=hash)
+
+    _cache: dict[Any, EndT] = dataclasses.field(repr=False, init=False, default_factory=dict)
```

Comment:
```suggestion
    hash_function: Callable[[StartT], HashT] = dataclasses.field(default=hash)

    _cache: dict[HashT, EndT] = dataclasses.field(repr=False, init=False, default_factory=dict)
```

No reason to use `Any` here, these types have to match. Just requires another unbounded TypeVar above.

Reply:
Done. I had add a type ignore as mypy is (in my understanding of the typing rules in python) wrongly complaining about the default `hash` function.